Responsive widget tweaks

sfc-gh-jgarcia opened the pull request:
## Describe your changes

This PR fixes a couple issues reported by @jrieke:
1. Truncate long `st.selectbox` and `st.multiselect` options in small resolutions (mainly when embedding apps in reduced containers);
2. Don't show the `Press enter to apply` text inside inputs when they're smaller than `120px`.

**Before:**

![Untitled](https://github.com/streamlit/streamlit/assets/103376966/202d055d-9289-4752-8a4b-8bdabf139fac)

<img width="452" alt="Untitled (1)" src="https://github.com/streamlit/streamlit/assets/103376966/1d7e9731-fd9b-4f91-a543-16685fee2f23">

**After:**

<img width="272" alt="Screenshot 2023-08-09 at 11 43 17 AM" src="https://github.com/streamlit/streamlit/assets/103376966/db1bda93-ba18-4df3-8b13-3b0f336fa821">

<img width="624" alt="Screenshot 2023-08-09 at 11 57 29 AM" src="https://github.com/streamlit/streamlit/assets/103376966/0e3b73be-3cff-4936-b994-723a65a407fe">

## GitHub Issue Link (if applicable)

* https://www.notion.so/snowflake-corp/43c412baa1954b7fb1dbe9e645ee321e?pvs=4
* https://www.notion.so/snowflake-corp/Long-selectbox-options-overlap-b81808f147d342609a769a837f3508a6?pvs=4

## Testing Plan

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-jgarcia committed:
Hide dirty input messages on small sizes

sfc-gh-jgarcia committed:
Truncate lenghty options across all screen resolutions

sfc-gh-jgarcia committed:
Improve size checking mechanism

sfc-gh-jgarcia committed:
Improve isVisible checking to ensure tests pass

sfc-gh-jgarcia committed:
Fix linting errors

kmcgrady commented:
Hey @sfc-gh-jgarcia there seems to be some issues here.

```python
import streamlit as st

cols = st.columns(8)
cols[0].number_input("Something")
```

See image

<img width="203" alt="Screenshot 2023-08-11 at 4 01 49 PM" src="https://github.com/streamlit/streamlit/assets/69432/27dc050b-7aff-43ed-93e5-602278d8bb2b">

Have you thought about [container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_container_queries) by chance? I think the caveat is that Safari doesn't support it in 15 (2 major versions) that being said, a new version of safari is around the corner AND, the worst case for this scenario is fine (the less than desired design for this specific use case).

sfc-gh-jgarcia commented:
Hi @kmcgrady! Thanks for the feedback. Strange, I'm unable to replicate your scenario on the `responsive-widget-tweaks` branch, this is what I'm getting 👇 

https://github.com/streamlit/streamlit/assets/103376966/ee0e266f-68e3-4970-875b-beb3d4440780

Besides that, I tried refactoring things to use container queries, but I couldn't get it to work neither using emotion nor through plain ol' CSS. Are we doing any transpiling/specific browser support that might be striping off these `@container` queries at build time?

kmcgrady commented:
Start with the following code:
```python
import streamlit as st

num_cols = st.number_input("Number of Columns", 2, 8, 2, 1)
cols = st.columns(num_cols)
cols[0].number_input("Something")
```

1. Type in the Something number input. Note the text.
1. Increase the number of columns to 8.
1. Start typing into the Something number Input. 

You'll notice the text appearing to the left.

kmcgrady commented:
For the container queries, you'll need to set the [container-type](https://developer.mozilla.org/en-US/docs/Web/CSS/container-type) CSS on the containers for reference. This may be the widget container itself (text input text area) or the styled component container (chat input and number input). I set it to "inline-size" and it worked fine.

You might also want to consider [container names](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_container_queries#naming_containment_contexts) as that's more semantic in meaning rather than relying on the nearest ancestor.

sfc-gh-jgarcia committed:
Use container queries to show/hide the instructions

sfc-gh-jgarcia commented:
Thanks for the explanation, @kmcgrady! Managed to get it working after your help in [9e52348](https://github.com/streamlit/streamlit/pull/7145/commits/9e5234824ab96c6663ff95c70a671ddfdba7966f). Setting the container type on `.StyledElementContainer` seems to work for all inputs and chat elements

sfc-gh-jgarcia committed:
Fix linting error

sfc-gh-jgarcia committed:
Merge branch 'develop' into responsive-widget-tweaks

sfc-gh-jgarcia committed:
Update playwright snapshots

sfc-gh-jgarcia committed:
Merge develop branch

sfc-gh-jgarcia committed:
Remove offending snapshots

sfc-gh-jgarcia committed:
Add back new snapshots

sfc-gh-jgarcia committed:
Remove snapshots from correct folder

sfc-gh-jgarcia committed:
Update snapshots

sfc-gh-jgarcia committed:
Delete wrongly generated folder

sfc-gh-jgarcia committed:
Delete snapshots

sfc-gh-jgarcia committed:
Add all snapshots back

lukasmasuch committed:
Merge remote-tracking branch 'remote/develop' into responsive-widget-tweaks

lukasmasuch committed:
Update snapshots

lukasmasuch commented on frontend/lib/src/components/core/Block/styled-components.ts:
```diff
@@ -57,6 +57,7 @@ export interface StyledElementContainerProps {
 const GLOBAL_ELEMENTS = ["balloons", "snow", "chatInput"]
 export const StyledElementContainer = styled.div<StyledElementContainerProps>(
   ({ theme, isStale, width, elementType }) => ({
+    containerType: "inline-size",
```
Do we need to apply this for all element containers? Based on the changes in the snapshot, it seems to have some small impact on a lot of elements in Chrome

sfc-gh-jgarcia replied:
I can try and add it on the widget container itself, but as @kmcgrady points out above, for number and chat inputs, that won't be possible

lukasmasuch committed:
Delete all snapshots

lukasmasuch committed:
Update snapshots

lukasmasuch reviewed (commented):
LGTM 👍  But maybe double-check if these subtle changes in Chrome are fine. The `containerType: "inline-size"` seems to have some impact on a couple of elements.

lukasmasuch commented:
As a bonus it would be awesome to have one test (either e2e or jest/rtl) to confirm the behavior of 2. However, it could be a bit more tricky :( 
For e2e test, you could put a number input in a very small column and call [fill](https://playwright.dev/python/docs/input#text-input) on it and then check if the instruction element appears or screenshot. But we don't have a playwright number_input test yet, so it might need a new one until we have migrated the old one.  For jest/rtl you can maybe simulate this by providing a width < 120 to the number input component and checking `display: "none"`.

lukasmasuch commented:
I just saw in NumberInput that we are doing something very similar already for the increment/decrement actions by just not rendering the sub-component on a specific width:

https://github.com/streamlit/streamlit/blob/0d065cd307f2e908c3ed76b22ef67049995d8765/frontend/lib/src/components/widgets/NumberInput/NumberInput.tsx#L416

Wouldn't it be possible to do the same also in the case of 2 for `StyledInstructionsContainer` without even using the CSS logic?

sfc-gh-jgarcia committed:
Remove previous implementation with container queries

sfc-gh-jgarcia committed:
New implementation using widget width

sfc-gh-jgarcia committed:
Add unit tests

sfc-gh-jgarcia commented:
> I just saw in NumberInput that we are doing something very similar already for the increment/decrement actions by just not rendering the sub-component on a specific width

You're right, I made that update a while back and totally forgot about it lol. Just updated it to work that way [here](https://github.com/streamlit/streamlit/pull/7145/commits/378bcbf6bd049fc81b33eb3cc9b861c2ed29d492), and added unit tests for all but the chat input component [here](https://github.com/streamlit/streamlit/pull/7145/commits/ed28e43550df5e1e6f177c283269b17e3165c425). I'm not sure when instructions are shown on `st.chat_input` (they aren't showing as you type), so I'm not really sure how to test them.

sfc-gh-jgarcia committed:
Delete snapshots

sfc-gh-jgarcia committed:
Add back snapshots

sfc-gh-lmasuch commented on frontend/lib/src/components/widgets/ChatInput/ChatInput.tsx:
```diff
@@ -218,16 +219,19 @@ function ChatInput({ width, element, widgetMgr }: Props): React.ReactElement {
-              inForm={false}
-            />
-          </StyledInputInstructionsContainer>
+          {/* Hide the "Please enter to apply" text in small widget sizes */}
```
nit: Chat input doesn't use the `Please enter to apply`, it is only using this if the max number of chars is limited. So the comment is bit misleading, but we probably still want to hide it for the `maxChars` case.

sfc-gh-jgarcia replied:
Gotcha! Fixed it in [4eab899](https://github.com/streamlit/streamlit/pull/7145/commits/4eab8994db300b1a23dc47972f86b118a2f516de)

lukasmasuch reviewed (approved):
LGTM 👍

sfc-gh-jgarcia committed:
Fix misleading comment on chat input
